### ISSUE TITLE
sortable-objects rerenders #each after dragging

### ISSUE THREAD

**Issue by AndyPyeSC** [mharris717/ember-drag-drop#20]
I'm trying to add a sortable list to an app I'm building, but have found that after a drag-drop operation, the list of items will usually be rendered in the wrong order. The underlying list is in the correct order.

I thought this was an issue with my code, but after copying in the template, controller and styles from the sortable example, I'm still seeing exactly the same behaviour. When I view your example online, it works perfectly! What could I be missing?

**Comment by dgavey** [comment]
Hmm, not sure.  I had this issue the other day when I was implementing it, and I was passing the wrong variable to content= in the draggable-object component.

Maybe if you could post the template code for what you got, I might be able to figure it out.

**Comment by AndyPyeSC** [comment]
I got to the point where I literally copied the template, controller and styles from the demo into my app, and that still didn't work...

I've since changed to a different package ([ember-sortable](https://github.com/jgwhite/ember-sortable)), which initially showed the same behaviour until I realised that I wasn't updating the original array in my "order changed" handler. I'm not sure this was the same problem as I experienced with yours though. Sorry I can't be any more help!

**Comment by dgavey** [comment]
Well, in this addon, the library updates the array for you so you shouldn't have to do it yourself.  Are you using ember-data?  I haven't tested with that yet.  I have tested with ember-model and plain arrays though.

**Comment by AndyPyeSC** [comment]
Yes, ember-sortable uses the "Data Down, Actions Up" pattern. I started off using ember-data (I was displaying the items in a hasMany relationship), but as I said, I copied the example code in just to see if that works, and it still didn't (i.e. the array was a plain array as a property on the controller).

**Comment by dgavey** [comment]
Well, in my experience a 'plain' array will not re-render on a change.  It would have to be an Ember array that you past in.  This should have worked with Ember data though.  This probably isn't the issue though.  Sorry it didn't work for you, wish I had an answer.  Of course a failing js-bin would be super helpful, but I understand that's probably not possible.

**Comment by AndyPyeSC** [comment]
No worries, sorry I can't provide a demonstration!

**Comment by gnazarkin** [comment]
Having the exact problem described here. The underlying array is sorted properly but the displayed list in the sortable-objects component is displayed incorrectly

**Comment by dgavey** [comment]
What is the structure of the array you are sending to it? POJO Ember Data? Ember Object?

**Comment by gnazarkin** [comment]
Ember Data. This is how I have it implemented right now
```
{{#sortable-objects sortableObjectList=model.fields sortEndAction='sortEndAction' enableSort=true}}
  {{#each model.fields as |item|}}
    {{#draggable-object content=item isSortable=true}}
      {{item.name}}
    {{/draggable-object}}
  {{/each}}
{{/sortable-objects}}

**Comment by dgavey** [comment]
Well your template looks ok.  I'm getting the picture it's a problem with Ember Data objects.  Gonna try and make a failing case later.  What is the model definition for model.fields? 

Also visually as you drag an element over another element the elements nodes should swap. So visually you should see the element change position as soon as your overtop of a new element.  You can look at the demo to see what I mean.  Do you see that visual change?

**Comment by gnazarkin** [comment]
It's a hasMany relationship. model hasMany fields and field belongsTo model. And the visuals are working fine with the changes as described

**Comment by dgavey** [comment]
Ok, thanks for the info, I'll try and get a failing case using Ember Data.  Of course, if you are able to make me a failing JSBin that would be awesome!

**Comment by dgavey** [comment]
Sorry, got busy. I found time tonight to put an example together using Ember Data and my simple example seems to work fine.  I'm still at a complete loss to figure out what the issue is.  If there is any way I could play with the code that is having the issue that would be helpful.  Otherwise I guess I need a failing case to work from.

**Comment by gnazarkin** [comment]
If you want I can screenshare and show you my code/implementation. Shoot me an email at gleb.nazarkin@gmail.com and we can figure something out

**Comment by dgavey** [comment]
I've finally nailed down where the issue is.  It seems the problem affects all Glimmer based releases of Ember 1.13 and up.  I'm still working on a fix, hopefully soon.  It has to do with my use of swapping nodes in the DOM.  Glimmer releases of Ember get confused, previous releases are fine.

**Comment by trumb1mj** [comment]
I copied the exact code from the demo and am seeing the same behavior described in this issue.

**Comment by gnazarkin** [comment]
@trumb1mj the fix is on the develop branch now. I'm using that branch with no problems

**Comment by dgavey** [comment]
An Ember version 1.13.2 plus compatible version of sortable objects has been published under version 0.3.0 . Integration tests have been written as well. So hopefully I'll know of failing cases in new ember versions first.
